internal/provider/cluster_resource.go

```diff
@@ -716,23 +701,64 @@ func (r *clusterResource) Update(
 			return
 		}
 
-		upgradeStatus := client.CLUSTERUPGRADESTATUSTYPE_MAJOR_UPGRADE_RUNNING
-		cmp := compareCrdbVersions(stateVersion, planVersion, &resp.Diagnostics)
-		if cmp < 0 {
-			// Make sure we're rolling back to the previous version.
-			if cmp < -1 {
-				resp.Diagnostics.AddError("Invalid rollback version", "Can only roll back to the previous version.")
+		// Next, if we are upgrading, it must be a valid upgrade from the
+		// current version, according to the ListMajorClusterVersions API.
+		if upgrading, err := isUpgrade(stateVersion, planVersion); upgrading {
+			var currentVersionInfo client.ClusterMajorVersion
+			for _, v := range apiResp.Versions {
+				if v.Version == stateVersion {
+					currentVersionInfo = v
+					break
+				}
+			}
+
+			var validUpgrade bool
+			for _, v := range currentVersionInfo.AllowedUpgrades {
+				if v == planVersion {
+					validUpgrade = true
+					break
+				}
+			}
+			if !validUpgrade {
+				resp.Diagnostics.AddError("Invalid CockroachDB version",
+					fmt.Sprintf(
+						"Cannot change major version to '%s'. Valid upgrade versions include [%s]",
+						planVersion,
+						strings.Join(currentVersionInfo.AllowedUpgrades, "|")))
 				return
 			}
-			upgradeStatus = client.CLUSTERUPGRADESTATUSTYPE_ROLLBACK_RUNNING
-		} else if cmp > 1 {
-			resp.Diagnostics.AddError("Invalid upgrade version", "Can't skip versions. Upgrades must be performed one version at a time.")
+			if IsKnown(plan.UpgradeStatus) && plan.UpgradeStatus.ValueString() != string(client.CLUSTERUPGRADESTATUSTYPE_MAJOR_UPGRADE_RUNNING) {
+				resp.Diagnostics.AddError("Invalid upgrade_status",
+					fmt.Sprintf(
+						"cockroach_version ('%s') would upgrade cluster, but upgrade_status was '%s'. "+
+							"Remove upgrade_status when upgrading via cockroach_version.",
+						planVersion,
```

Comment:
I was somewhat surprised by this comment as it seems to be describing how upgrade_status in the api is used which should be an internal concern.  Then I was thinking that maybe I was misremembering how upgrades work in terraform so I had to review.  Note [here](https://github.com/cockroachdb/terraform-provider-cockroach/blob/main/internal/provider/cluster_resource.go#L267-L270) that `upgrade_status` is marked as `Computed: true` and does not include `Optional: true,`.  This means they can't actually set this value in the plan.   IIRC, version upgrades in terraform just rely on the settable `cockroach_version` attribute.  if the version in the plan is later than the state, we try to upgrade.  If the version in the plan is earlier, we try to rollback.  This flow relies on the 72 hour auto finalization.  In order to finalize more quickly the `cockroach_finalize_version_upgrade` resource can be used.    Maybe I've misunderstood what's going on here.  Let's chat about it.  Also, side note, I've wanted to redesign how finalization works for some time now.  Maybe now is a good time to do it.  Or it could wait.

Reply:
Ah, yeah, this was a holdover from my earlier misunderstanding about the `upgrade_status` field -- thanks for catching, I will remove this error checking block, which is moot, I think